packages/devtools_app/lib/src/shared/diagnostics/inspector_service.dart

```diff
@@ -418,12 +418,53 @@ class InspectorService extends InspectorServiceBase {
     return false;
   }
 
+  /// Ensures that addPubRootDirectories, getPubRootDirectories, and
+  /// removePubRootDirectories have all registered as services
+  /// before completing.
+  Future<void> waitUntilPubRootDirectoriesServicesAvailable() {
+    final futureListener = Completer();
+    bool addAvailable = false;
+    bool getAvailable = false;
+    bool removeAvailable = false;
+    void checkIfComplete() {
+      if (addAvailable && getAvailable && removeAvailable) {
+        futureListener.complete();
+      }
+    }
+
+    serviceConnection.serviceManager
+        .registeredServiceListenable(
+            WidgetInspectorServiceExtensions.addPubRootDirectories.name)
+        .addListener(() {
```

Comment:
these listeners need to be removed once we are finished with them. We also need to make sure the value of the `registeredServiceListenable(WidgetInspectorServiceExtensions.addPubRootDirectories.name)` listenable is true before we say that it is available. The value could come back as false in which case the service would not be available.

Reply:
example: https://github.com/flutter/devtools/blob/master/packages/devtools_app_shared/lib/src/service/connected_app.dart#L157-L171